### DISCUSSION TITLE
What's the best way to make an extensible format with Cap'n Proto?

### DISCUSSION THREAD

**Discussion by mikehearn** [capnproto/capnproto#2004] [General]
In protocol buffers it's possible to let third parties extend protocols in a relatively uncoordinated way. You can reserve a range of tag values in your root struct and then have some central authority dish them out, for example, or you can implement the moral equivalent of Map<String, Bytes> with embedded protobufs.

In Cap'n Proto it's less obvious how to do this. Tag values must be sequential. Embedding binary encodings is never ideal and maybe loses the nice properties of the format.

Perhaps the best approach is something like:

```
struct Extensible {
    extensions @0 :Map(Text, Any)
    # Map of reverse DNS names (com.example.foo) to other objects.
}
```

What do you think?

**Comment by kentonv** [comment]
In protobuf, "extensions" made a lot of sense because they fit into the wire format so elegantly. In Cap'n Proto, of course, we don't use tags, so there's not such a great fit. On the other hand, `AnyPointer` turns out to be a very natural fit for Cap'n Proto, so I went that way.

When you really need to support multiple extensions in the same message, a map (or list of key/value pairs) mapping some sort of type IDs (maybe `Text`, maybe `UInt64`) to `AnyPointer` makes sense, yes.

But in my experience, a lot of cases don't actually call for multiple extensions. If only one extension is expected at a time, then you can simply use `AnyPointer`, or better yet, generics.

```
struct Extensible(T) {
  extension @0 :T;
}
```

It all depends on the use case, though.

**Comment by mikehearn** [comment]
The use case here is a data format that's meant to be open and live for a
long time, so, to be robust against change it'd be nice to have a way for
others to extend it. This is a common feature of really long lived open
formats.
